File: components/src/main/java/net/yakclient/graphics/components/TextSpecProps.java
package net.yakclient.graphics.components;

import net.yakclient.graphics.api.gui.GuiProperties;
import net.yakclient.graphics.api.gui.IllegalPropertyException;
import net.yakclient.graphics.api.gui.SpecifiedProperties;
import net.yakclient.graphics.util.ColorCodes;
import net.yakclient.graphics.util.RGBColor;
import net.yakclient.graphics.util.YakFont;
import net.yakclient.graphics.util.YakFontFactory;

import java.util.Optional;

public class TextSpecProps extends SpecifiedProperties {
    public TextSpecProps(GuiProperties properties) {
        super(properties);
    }

    public String getValue() {
        return this.<String>get("value").orElseThrow(()-> new IllegalPropertyException("value"));
    }

    public double getX() {
        return this.<Double>get("x").orElseThrow(()->new IllegalPropertyException("x"));
    }

    public double getY() {
        return this.<Double>get("y").orElseThrow(()->new IllegalPropertyException("y"));
    }

    public RGBColor getColor() {
        return this.<RGBColor>get("color").or(()-> Optional.of(ColorCodes.WHITE)).get();
    }

    public YakFont getFont() {
        return this.<YakFont>get("font").or(()-> Optional.of(YakFontFactory.create().buildFont())).get();
    }
}

File: components/src/main/kotlin/net/yakclient/graphics/components/TextSpecProps.kt
package net.yakclient.graphics.components

import net.yakclient.graphics.api.GuiPropertiesMap
import net.yakclient.graphics.api.IllegalPropertyException
import net.yakclient.graphics.api.SpecifiedProperties
import net.yakclient.graphics.util.Color
import net.yakclient.graphics.util.ColorCodes
import net.yakclient.graphics.util.TextureFont
import net.yakclient.graphics.util.FontFactory.fontOf
import java.util.*

public class TextSpecProps(properties: GuiPropertiesMap?) : SpecifiedProperties(properties!!) {
    public val value: String
        get() = this.get<String>("value").orElseThrow { IllegalPropertyException("value") }
    public val x: Double
        get() = this.get<Double>("x").orElseThrow { IllegalPropertyException("x") }
    public val y: Double
        get() = this.get<Double>("y").orElseThrow { IllegalPropertyException("y") }
    public val color: Color
        get() = this.get<Color?>("color").or { Optional.of(ColorCodes.WHITE) }
            .get()
    public val font: TextureFont
        get() = this.get<TextureFont>("font").or { Optional.of(fontOf()) }.get()
}